plugin/kubernetes/metadata.go

```diff
@@ -3,12 +3,14 @@ package kubernetes
 import (
 	"context"
 
+	"github.com/coredns/coredns/plugin"
 	"github.com/coredns/coredns/plugin/metadata"
 	"github.com/coredns/coredns/request"
 )
 
 // Metadata implements the metadata.Provider interface.
 func (k *Kubernetes) Metadata(ctx context.Context, state request.Request) context.Context {
+	state.Zone = plugin.Zones(k.Zones).Matches(state.Name())
```

Comment:
Does this mutate the state in some undesired way?

Also, can we have a unit test for the bug and show that this fixes it?

Reply:
> Does this mutate the state in some undesired way? 

Yes, see the second commit...

Reply:
A more graceful way to handle this would be to have parse accept the matched zone, instead of pulling it from state.

That the kubernetes plugin modifies state.Zone at all is potentially problematic for fallthrough cases.

Reply:
> Also, can we have a unit test for the bug and show that this fixes it?

Added a test and verified the unit test fails when without the fix.

Reply:
can you go ahead with the "more graceful way"? Is there a blocker for that?

Reply:
Sure, no blocker - it just touches more code.